src/fsharp/IlxGen.fs

```diff
@@ -2820,11 +2820,15 @@ and GenNewArraySimple cenv cgbuf eenv (elems, elemTy, m) sequel =
     let ilElemTy = GenType cenv.amap m eenv.tyenv elemTy
     let ilArrTy = mkILArr1DTy ilElemTy
 
-    CG.EmitInstrs cgbuf (pop 0) (Push [ilArrTy]) [ (AI_ldc (DT_I4, ILConst.I4 (elems.Length))); I_newarr (ILArrayShape.SingleDimensional, ilElemTy) ]
-    elems |> List.iteri (fun i e ->         
-        CG.EmitInstrs cgbuf (pop 0) (Push [ilArrTy; cenv.g.ilg.typ_Int32]) [ AI_dup; (AI_ldc (DT_I4, ILConst.I4 i)) ]
-        GenExpr cenv cgbuf eenv SPSuppress e Continue      
-        CG.EmitInstr cgbuf (pop 3) Push0 (I_stelem_any (ILArrayShape.SingleDimensional, ilElemTy)))
+    if elems.Length = 0 then
+        let emptyCallIns = mkNormalCall (mkILMethSpecInTy (cenv.g.ilg.typ_Array, ILCallingConv.Static, "Empty", [], mkILArr1DTy (mkILTyvarTy 0us), [ilElemTy]))
```

Comment:
Should `"Empty"` be defined somewhere in `ILGlobals`? It seemed to specific for this scenario so I didn't bother.

Reply:
Array.Empty isn't available before net framework 4.6 so I expect this needs a type lookup and a fallback

Reply:
Well, that is true, I misunderstood this comment https://github.com/dotnet/fsharp/issues/7595#issuecomment-533811045. The fallback could simply be a regular allocation, but I am not sure how to do that type lookup depending on the TFM of user code.

Reply:
On the other hand, according to [this](https://github.com/dotnet/fsharp/blob/main/docs/fsharp-core-notes.md#compatibility-table), F# 5.0 requires FSharp.Core 5.0, which only supports netstandard2.0. So if this change is released as, say, F# 5.1 (even though it isn't a language feature), then it doesn't matter because you cannot (should not?) target net46 with F# 5.0+?

Reply:
It would be best to have it in `TcGlobals`. Then when using it in IlxGen, add a check to see if we can deref the method as it may not be available.

Reply:
> check to see if we can deref the method

I don't really know what that is. Can you please show me an example of that somewhere?

Reply:
I was hoping this would tell me whether the method is available, but I'm just getting None
```fsharp
let key = ValLinkageFullKey ({ MemberParentMangledName = Some TcGlobals.tname_Array; MemberIsOverride = false; LogicalName = "Empty"; TotalArgCount = 0 }, None)
let vref = mkNonLocalValRef cenv.g.system_Array_tcref.nlr key
let _d = vref.TryDeref
```

Reply:
`None` means it is not available. There is `TryDeref` and `CanDeref` - both are basically the same, just that `TryDeref` returns an option.

Here is an example of checking to see if the `IsReadOnlyAttribute` exists.
```fsharp
let GenReadOnlyAttributeIfNecessary (g: TcGlobals) ty =
    let add = isInByrefTy g ty && g.attrib_IsReadOnlyAttribute.TyconRef.CanDeref
    if add then
        let attr = mkILCustomAttribute g.ilg (g.attrib_IsReadOnlyAttribute.TypeRef, [], [], [])
        Some attr
    else
        None
```

Reply:
But it should be available (targetting .NET Core), so I am obviously doing something wrong. The problem is I don't see an example of getting hold of a vref for a static method on some System class in TcGlobals. It's either classes (attributes) or operators and intrinsics from FSharp.Core (`makeIntrinsicValRef` and then`ValRefForIntrinsic`). There's already `system_Array_tcref  = findSysTyconRef sys "Array"`, but I need a way to get to its methods.